featurebyte/query_graph/sql/feature_compute.py

```diff
@@ -154,18 +164,112 @@ def iter_aggregators(self) -> Iterable[AggregatorType]:
     def add_aggregation_spec(
         self,
         aggregation_spec: AggregationSpec,
+        node_name: str,
     ) -> None:
         """Add AggregationSpec to be incorporated when generating SQL
 
         Parameters
         ----------
         aggregation_spec : AggregationSpec
             Aggregation specification
+        node_name: str
+            Feature node name associated with the aggregation specification
         """
+        self.update_feature_entity_lookup_steps(aggregation_spec, node_name)
         key = aggregation_spec.aggregation_type
         aggregator = self.aggregators[key]
         aggregator.update(aggregation_spec)  # type: ignore
 
+    def update_feature_entity_lookup_steps(self, agg_spec: AggregationSpec, node_name: str) -> None:
+        """
+        Check if any entity lookup steps are required for the aggregation spec and if so update its
+        serving_names attribute in place. Also add the required lookup steps into the plan.
+
+        Parameters
+        ----------
+        agg_spec: AggregationSpec
+            Aggregation specification
+        node_name: str
+            Feature node name associated with the aggregation specification
+        """
+        entity_relationships_context = (
+            self.parent_serving_preparation.entity_relationships_context
+            if self.parent_serving_preparation is not None
+            else None
+        )
+        if entity_relationships_context is None:
+            return
+
+        parent_entity_columns = self._get_parent_entity_columns(
+            entity_relationships_context=entity_relationships_context,
+            node_info=entity_relationships_context.get_node_info(node_name),
+            agg_spec=agg_spec,
+        )
+
+        # Add to overall feature entity lookup steps
+        for entity_column in parent_entity_columns.values():
+            if entity_column.relationship_info_id is not None:
+                entity_lookup_step = (
+                    entity_relationships_context.entity_lookup_step_creator.get_entity_lookup_step(
+                        relationship_info_id=entity_column.relationship_info_id,
+                        child_serving_name_override=entity_column.child_serving_name,
+                        parent_serving_name_override=entity_column.serving_name,
+                    )
+                )
+                self.add_feature_entity_lookup_step(entity_lookup_step)
+
+        # Get updated serving names in aggregation spec
+        assert agg_spec.entity_ids is not None
+        new_serving_names = []
+        for entity_id, serving_name in zip(agg_spec.entity_ids, agg_spec.serving_names):
+            if entity_id in parent_entity_columns:
+                new_serving_name = parent_entity_columns[entity_id].serving_name
+            else:
+                new_serving_name = serving_name
+            new_serving_names.append(new_serving_name)
```

Comment:
Just to check my understanding here, the main purpose of doing this is to use the frozen relationship (as the entity here refers to the relationship table one but not the user provided parent entity)?

Reply:
Yes, the entity columns here are a result of looking up relation tables, not the ones provided by the user.